packages/core/index.d.ts

```diff
@@ -1,57 +1,72 @@
 import {
-  Callback,
-  Context,
-  Handler
+  Context as LambdaContext,
+  Handler as LambdaHandler
 } from 'aws-lambda'
 
-declare type EventType<T, C> =
-  T extends (event: infer EventArgType, context: C, callback: Callback<any>) => void ? EventArgType :
-  T extends (event: infer EventArgType, context: C) => Promise<any> ? EventArgType :
-  never;
+declare type PluginHook = () => void;
+declare type PluginHookWithMiddlewareName = (middlewareName: string) => void;
 
-declare type HandlerReturnType<T, C> =
-  T extends (event: any, context: C) => Promise<infer RetType> ? RetType :
-  T extends (event: any, context: C, callback: Callback<infer RetType>) => void ? RetType :
-  never;
+declare type PluginObject = {
+  beforePrefetch?: PluginHook;
+  requestStart?: PluginHook;
+  beforeMiddleware?: PluginHookWithMiddlewareName;
+  afterMiddleware?: PluginHookWithMiddlewareName;
+  beforeHandler?: PluginHook;
+  afterHandler?: PluginHook;
+  requestEnd?: PluginHook;
```

Comment:
This can be a promise, not sure if that needs a different type

Reply:
Great point, yeah we will need the method to return a promise...